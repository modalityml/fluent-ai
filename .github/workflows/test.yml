name: Test

on:
  push:
    branches:
      - main

  pull_request:
    branches:
      - main

jobs:
  test:
    runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest]
      fail-fast: false

    steps:
      - id: checkout
        name: Checkout
        uses: actions/checkout@v3
      - id: setup-bun
        name: Setup Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest
      - id: install-deps
        name: Install dependencies
        run: |
          bun install
      - id: tsc
        run: bunx tsc
        working-directory: packages/fluent-ai
      - id: build
        name: Run build
        run: |
          bun run build
        working-directory: packages/fluent-ai
      - id: test
        name: Run test
        run: |
          bun test
        working-directory: packages/fluent-ai
