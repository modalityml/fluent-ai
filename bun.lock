{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "fluent-ai",
      "dependencies": {
        "eventsource-parser": "^3.0.6",
        "partial-json": "^0.1.7",
      },
      "devDependencies": {
        "@types/bun": "1.3.0",
        "bun-plugin-dts": "^0.3.0",
        "prettier": "^3.6.2",
        "typescript": "^5.9.3",
        "zod": "^4.1.12",
      },
      "peerDependencies": {
        "zod": "^4",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.3.0", "", { "dependencies": { "bun-types": "1.3.0" } }, "sha512-+lAGCYjXjip2qY375xX/scJeVRmZ5cY0wyHYyCYxNcdEXrQ4AOe3gACgd4iQ8ksOslJtW4VNxBJ8llUwc3a6AA=="],

    "@types/node": ["@types/node@22.13.5", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-+lTU0PxZXn0Dr1NBtC7Y8cR21AJr87dLLU953CWA6pMxxv/UDc7jYAY90upcrie1nRcD6XNG5HOYEDtgW5TxAg=="],

    "@types/react": ["@types/react@19.2.2", "", { "dependencies": { "csstype": "^3.0.2" } }, "sha512-6mDvHUFSjyT2B2yeNx2nUgMxh9LtOWvkhIU3uePn2I2oyNymUAX1NIsdgviM4CH+JSrp2D2hsMvJOkxY+0wNRA=="],

    "ansi-regex": ["ansi-regex@5.0.1", "", {}, "sha512-quJQXlTSUGL2LH9SUXo8VwsY4soanhgo6LNSm84E1LBcE8s3O0wpdiRzyR9z/ZZJMlMWv37qOOb9pdJlMUEKFQ=="],

    "ansi-styles": ["ansi-styles@4.3.0", "", { "dependencies": { "color-convert": "^2.0.1" } }, "sha512-zbB9rCJAT1rbjiVDb2hqKFHNYLxgtk8NURxZ3IZwD3F6NtxbXZQCnnSi1Lkx+IDohdPlFp222wVALIheZJQSEg=="],

    "bun-plugin-dts": ["bun-plugin-dts@0.3.0", "", { "dependencies": { "common-path-prefix": "^3.0.0", "dts-bundle-generator": "^9.5.1", "get-tsconfig": "^4.8.1" } }, "sha512-QpiAOKfPcdOToxySOqRY8FwL+brTvyXEHWzrSCRKt4Pv7Z4pnUrhK9tFtM7Ndm7ED09B/0cGXnHJKqmekr/ERw=="],

    "bun-types": ["bun-types@1.3.0", "", { "dependencies": { "@types/node": "*" }, "peerDependencies": { "@types/react": "^19" } }, "sha512-u8X0thhx+yJ0KmkxuEo9HAtdfgCBaM/aI9K90VQcQioAmkVp3SG3FkwWGibUFz3WdXAdcsqOcbU40lK7tbHdkQ=="],

    "cliui": ["cliui@8.0.1", "", { "dependencies": { "string-width": "^4.2.0", "strip-ansi": "^6.0.1", "wrap-ansi": "^7.0.0" } }, "sha512-BSeNnyus75C4//NQ9gQt1/csTXyo/8Sb+afLAkzAptFuMsod9HFokGNudZpi/oQV73hnVK+sR+5PVRMd+Dr7YQ=="],

    "color-convert": ["color-convert@2.0.1", "", { "dependencies": { "color-name": "~1.1.4" } }, "sha512-RRECPsj7iu/xb5oKYcsFHSppFNnsj/52OVTRKb4zP5onXwVF3zVmmToNcOfGC+CRDpfK/U584fMg38ZHCaElKQ=="],

    "color-name": ["color-name@1.1.4", "", {}, "sha512-dOy+3AuW3a2wNbZHIuMZpTcgjGuLU/uBL/ubcZF9OXbDo8ff4O8yVp5Bf0efS8uEoYo5q4Fx7dY9OgQGXgAsQA=="],

    "common-path-prefix": ["common-path-prefix@3.0.0", "", {}, "sha512-QE33hToZseCH3jS0qN96O/bSh3kaw/h+Tq7ngyY9eWDUnTlTNUyqfqvCXioLe5Na5jFsL78ra/wuBU4iuEgd4w=="],

    "csstype": ["csstype@3.1.3", "", {}, "sha512-M1uQkMl8rQK/szD0LNhtqxIPLpimGm8sOBwU7lLnCpSbTyY3yeU1Vc7l4KT5zT4s/yOxHH5O7tIuuLOCnLADRw=="],

    "dts-bundle-generator": ["dts-bundle-generator@9.5.1", "", { "dependencies": { "typescript": ">=5.0.2", "yargs": "^17.6.0" }, "bin": { "dts-bundle-generator": "dist/bin/dts-bundle-generator.js" } }, "sha512-DxpJOb2FNnEyOzMkG11sxO2dmxPjthoVWxfKqWYJ/bI/rT1rvTMktF5EKjAYrRZu6Z6t3NhOUZ0sZ5ZXevOfbA=="],

    "emoji-regex": ["emoji-regex@8.0.0", "", {}, "sha512-MSjYzcWNOA0ewAHpz0MxpYFvwg6yjy1NG3xteoqz644VCo/RPgnr1/GGt+ic3iJTzQ8Eu3TdM14SawnVUmGE6A=="],

    "escalade": ["escalade@3.2.0", "", {}, "sha512-WUj2qlxaQtO4g6Pq5c29GTcWGDyd8itL8zTlipgECz3JesAiiOKotd8JU6otB3PACgG6xkJUyVhboMS+bje/jA=="],

    "eventsource-parser": ["eventsource-parser@3.0.6", "", {}, "sha512-Vo1ab+QXPzZ4tCa8SwIHJFaSzy4R6SHf7BY79rFBDf0idraZWAkYrDjDj8uWaSm3S2TK+hJ7/t1CEmZ7jXw+pg=="],

    "get-caller-file": ["get-caller-file@2.0.5", "", {}, "sha512-DyFP3BM/3YHTQOCUL/w0OZHR0lpKeGrxotcHWcqNEdnltqFwXVfhEBQ94eIo34AfQpo0rGki4cyIiftY06h2Fg=="],

    "get-tsconfig": ["get-tsconfig@4.10.0", "", { "dependencies": { "resolve-pkg-maps": "^1.0.0" } }, "sha512-kGzZ3LWWQcGIAmg6iWvXn0ei6WDtV26wzHRMwDSzmAbcXrTEXxHy6IehI6/4eT6VRKyMP1eF1VqwrVUmE/LR7A=="],

    "is-fullwidth-code-point": ["is-fullwidth-code-point@3.0.0", "", {}, "sha512-zymm5+u+sCsSWyD9qNaejV3DFvhCKclKdizYaJUuHA83RLjb7nSuGnddCHGv0hk+KY7BMAlsWeK4Ueg6EV6XQg=="],

    "partial-json": ["partial-json@0.1.7", "", {}, "sha512-Njv/59hHaokb/hRUjce3Hdv12wd60MtM9Z5Olmn+nehe0QDAsRtRbJPvJ0Z91TusF0SuZRIvnM+S4l6EIP8leA=="],

    "prettier": ["prettier@3.6.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-I7AIg5boAr5R0FFtJ6rCfD+LFsWHp81dolrFD8S79U9tb8Az2nGrJncnMSnys+bpQJfRUzqs9hnA81OAA3hCuQ=="],

    "require-directory": ["require-directory@2.1.1", "", {}, "sha512-fGxEI7+wsG9xrvdjsrlmL22OMTTiHRwAMroiEeMgq8gzoLC/PQr7RsRDSTLUg/bZAZtF+TVIkHc6/4RIKrui+Q=="],

    "resolve-pkg-maps": ["resolve-pkg-maps@1.0.0", "", {}, "sha512-seS2Tj26TBVOC2NIc2rOe2y2ZO7efxITtLZcGSOnHHNOQ7CkiUBfw0Iw2ck6xkIhPwLhKNLS8BO+hEpngQlqzw=="],

    "string-width": ["string-width@4.2.3", "", { "dependencies": { "emoji-regex": "^8.0.0", "is-fullwidth-code-point": "^3.0.0", "strip-ansi": "^6.0.1" } }, "sha512-wKyQRQpjJ0sIp62ErSZdGsjMJWsap5oRNihHhu6G7JVO/9jIB6UyevL+tXuOqrng8j/cxKTWyWUwvSTriiZz/g=="],

    "strip-ansi": ["strip-ansi@6.0.1", "", { "dependencies": { "ansi-regex": "^5.0.1" } }, "sha512-Y38VPSHcqkFrCpFnQ9vuSXmquuv5oXOKpGeT6aGrr3o3Gc9AlVa6JBfUSOCnbxGGZF+/0ooI7KrPuUSztUdU5A=="],

    "typescript": ["typescript@5.9.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-jl1vZzPDinLr9eUt3J/t7V6FgNEw9QjvBPdysz9KfQDD41fQrC2Y4vKQdiaUpFT4bXlb1RHhLpp8wtm6M5TgSw=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],

    "wrap-ansi": ["wrap-ansi@7.0.0", "", { "dependencies": { "ansi-styles": "^4.0.0", "string-width": "^4.1.0", "strip-ansi": "^6.0.0" } }, "sha512-YVGIj2kamLSTxw6NsZjoBxfSwsn0ycdesmc4p+Q21c5zPuZ1pl+NfxVdxPtdHvmNVOQ6XSYG4AUtyt/Fi7D16Q=="],

    "y18n": ["y18n@5.0.8", "", {}, "sha512-0pfFzegeDWJHJIAmTLRP2DwHjdF5s7jo9tuztdQxAhINCdvS+3nGINqPd00AphqJR/0LhANUS6/+7SCb98YOfA=="],

    "yargs": ["yargs@17.7.2", "", { "dependencies": { "cliui": "^8.0.1", "escalade": "^3.1.1", "get-caller-file": "^2.0.5", "require-directory": "^2.1.1", "string-width": "^4.2.3", "y18n": "^5.0.5", "yargs-parser": "^21.1.1" } }, "sha512-7dSzzRQ++CKnNI/krKnYRV7JKKPUXMEh61soaHKg9mrWEhzFWhFnxPxGl+69cD1Ou63C13NUPCnmIcrvqCuM6w=="],

    "yargs-parser": ["yargs-parser@21.1.1", "", {}, "sha512-tVpsJW7DdjecAiFpbIB1e3qxIQsE6NoPc5/eTdrbbIC4h0LVsWhnoa3g+m2HclBIujHzsxZ4VJVA+GUuc2/LBw=="],

    "zod": ["zod@4.1.12", "", {}, "sha512-JInaHOamG8pt5+Ey8kGmdcAcg3OL9reK8ltczgHTAwNhMys/6ThXHityHxVV2p3fkw/c+MAvBHFVYHFZDmjMCQ=="],

    "dts-bundle-generator/typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],
  }
}
