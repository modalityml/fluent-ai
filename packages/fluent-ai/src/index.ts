export * from "~/src/job/schema";
export * from "~/src/builder/chat";
export * from "~/src/builder/embedding";
export * from "~/src/builder/openrouter";
export * from "~/src/builder/voyage";
export * from "~/src/agent/agent";
export * from "~/src/agent/tool";
export * from "~/src/agent/repl";
