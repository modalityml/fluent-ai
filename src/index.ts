export * from "./jobs/job";
export * from "./jobs/load";
export * from "./jobs/utils";
export * from "./jobs/chat";
export * from "./jobs/embedding";
export * from "./jobs/image";
export * from "./jobs/models";

export * from "./providers/anthropic";
export * from "./providers/fal";
export * from "./providers/fireworks";
export * from "./providers/google";
export * from "./providers/luma";
export * from "./providers/ollama";
export * from "./providers/openai";
export * from "./providers/together";
export * from "./providers/voyageai";
