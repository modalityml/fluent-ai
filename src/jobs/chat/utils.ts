import { z } from "zod";
import { parse } from "partial-json";
import type { ChatToolSchema, Message } from "./schema";
import { ChatTool } from "./tool";

export function convertTools(tools: z.infer<typeof ChatToolSchema>[]) {
  return tools.map((tool) => ({
    type: "function",
    function: {
      name: tool.name,
      description: tool.description,
      parameters: z.toJSONSchema(tool.parameters),
    },
  }));
}

export function user(content: string) {
  return { role: "user", content: content } as Message;
}

export function assistant(content: string) {
  return { role: "assistant", content: content } as Message;
}

export function system(content: string) {
  return { role: "system", content: content } as Message;
}

export function audio() {
  throw new Error("Not implemented");
}

export function image() {
  throw new Error("Not implemented");
}

export function tool(name: string) {
  return new ChatTool(name);
}

export function text(result: any) {
  if (result.raw) {
    // output text
    return result.raw.choices[0].message.content;
  }
  if (result.choices[0].delta.content) {
    // chunk text
    return result.choices[0].delta.content;
  }
  return "";
}

export function object(result: any) {
  return JSON.parse(result.raw.choices[0].message.content);
}

export function partialParse(content: string) {
  if (content) {
    return parse(content);
  }
  return {};
}
