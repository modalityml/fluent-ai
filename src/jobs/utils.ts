// import { parse } from "partial-json";

export async function requestObject(request: Request) {
  return {
    url: request.url,
    method: request.method,
    headers: Object.fromEntries(request.headers as any),
    body: await request.text(),
  };
}

export function chunkText(chunk: any) {
  return chunk.choices[0].delta.content;
}

// TODO:
export function chunkObject(chunk: any) {}

// TODO:
export function chunkTools(chunk: any) {}
