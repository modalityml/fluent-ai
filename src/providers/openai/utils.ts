import z from "zod";
import type { ChatToolChoiceSchema, ChatToolSchema } from "~/jobs/chat";
import type { OpenAIToolDefinition, OpenAIToolChoice } from "./types";
export function convertTools(tools: z.infer<typeof ChatToolSchema>[]): OpenAIToolDefinition[] {
  return tools.map((tool) => ({
    type: "function",
    function: {
      name: tool.name,
      description: tool.description,
      parameters: z.toJSONSchema(tool.parameters),
    },
  }));
}


export function convertToolChoice(toolChoice: z.infer<typeof ChatToolChoiceSchema>): OpenAIToolChoice {
  if (!toolChoice) return "auto";
  
  switch (toolChoice.mode) {
    case "none":
      return "none";
    case "auto":
      return "auto";
    case "any":
      // If allowed_tools is specified and has exactly one tool, specify that tool
      if (toolChoice.allowed_tools && toolChoice.allowed_tools.length === 1) {
        return {
          type: "function",
          function: { name: toolChoice.allowed_tools[0] }
        };
      }
      // Otherwise, use auto
      return "auto";
    default:
      return "auto";
  }
}
