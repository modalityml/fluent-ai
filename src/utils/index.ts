import { parse } from "partial-json";

export function chunkText(chunk: any) {
  return chunk.choices[0].delta.content;
}

// TODO:
export function chunkObject(chunk: any) {}

// TODO:
export function chunkTools(chunk: any) {}
