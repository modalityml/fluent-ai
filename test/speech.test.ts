import { test, expect } from "bun:test";
import { load, elevenlabs, openai } from "../src";
import { requestObject } from "./utils";

function createJobs() {
  // prettier-ignore
  return [
    openai().speech('tts-1'),
    elevenlabs().speech("model"),
  ]
}

test("speech", async () => {
  for (const job of createJobs()) {
    expect(await requestObject(job.makeRequest())).toMatchSnapshot();
  }
});

test("dump", () => {
  for (const job of createJobs()) {
    expect(job.dump()).toMatchSnapshot();
  }
});

test("load", async () => {
  for (const job of createJobs()) {
    const req1 = await requestObject(load(job.dump()).makeRequest!());
    const req2 = await requestObject(job.makeRequest());
    expect(req1).toEqual(req2);
  }
});
