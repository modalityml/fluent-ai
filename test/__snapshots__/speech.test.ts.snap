// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`dump 1`] = `
{
  "cost": undefined,
  "input": {
    "model": "tts-1",
  },
  "options": {
    "apiKey": undefined,
  },
  "output": undefined,
  "performance": undefined,
  "provider": "openai",
  "type": "speech",
  "version": "0.3.0",
}
`;

exports[`dump 2`] = `
{
  "cost": undefined,
  "input": {
    "model": "model",
  },
  "options": {
    "apiKey": undefined,
  },
  "output": undefined,
  "performance": undefined,
  "provider": "elevenlabs",
  "type": "speech",
  "version": "0.3.0",
}
`;

exports[`speech 1`] = `
{
  "body": "",
  "headers": {},
  "method": "GET",
  "url": "https://api.openai.com/v1",
}
`;

exports[`speech 2`] = `
{
  "body": "",
  "headers": {},
  "method": "GET",
  "url": "https://api.elevenlabs.io/v1/text-to-speech",
}
`;
